marlowe-plutus/charli3/Language/Marlowe/Scripts/Charli3.hs

```diff
@@ -0,0 +1,217 @@
+            _ -> traceError "C3c" -- Script value or datum changed.
+        PV2.Datum charli3Datum =
+          case filter
+            (\TxInInfo{txInInfoResolved = TxOut{txOutValue}} -> valueOf txOutValue charli3CurrencySymbol charli3TokenName > 0)
+            txInfoReferenceInputs of
+            [TxInInfo{txInInfoResolved = TxOut{txOutDatum = OutputDatum datum}}] -> datum
+            [TxInInfo{txInInfoResolved = TxOut{txOutDatum = OutputDatumHash hash}}] ->
+              case findDatum hash txInfo of
+                Just datum -> datum
+                Nothing -> traceError "C3d" -- Charli3 datum not found.
+            _ -> traceError "C3e" -- Charli3 oracle feed not found.
+        charli3Oracle =
+          case C3.getPriceDatas . C3.OracleFeed $ charli3Datum of
+            [pd] -> C3.getPrice . C3.getPriceMap $ pd
+            _ -> traceError "C3f" -- Precisely one Charli3 price value not found.
+        (currencySymbol, roleName) = do
+          let valuesList = AssocMap.toList $ getValue ownValue
+          case valuesList of
+            [(possibleAdaSymbol, _), (currencySymbol, AssocMap.toList -> [(roleName, _)])]
+              | possibleAdaSymbol == adaSymbol -> (currencySymbol, roleName)
+            [(currencySymbol, AssocMap.toList -> [(roleName, _)]), _] -> (currencySymbol, roleName)
+            _ -> traceError "C3g" -- Precisely one role token not found.
+        marloweValidatorAddress = scriptHashAddress marloweValidatorHash
+        marloweInput = case find (\TxInInfo{txInInfoResolved} -> txOutAddress txInInfoResolved == marloweValidatorAddress) txInfoInputs of
+          Just input -> input
+          Nothing -> traceError "C3h" -- Marlowe script UTXO not found.
+        threadTokenOk = do
+          let marloweValue = txOutValue $ txInInfoResolved marloweInput
+          traceIfFalse "C3i" (valueOf marloweValue currencySymbol threadTokenName > 0) -- Marlowe thread token not found.
+        marloweRedeemerOk = do
+          let TxInInfo{txInInfoOutRef = marloweTxOutRef} = marloweInput
+              inputContentUsesRole (V1.IChoice (V1.ChoiceId choiceId (V1.Role role)) choiceNum) = choiceId == charli3ChoiceName && role == roleName && choiceNum == charli3Oracle
+              inputContentUsesRole _ = False
+              inputUsesRole (V1.Scripts.MerkleizedTxInput inputContent _) = inputContentUsesRole inputContent
+              inputUsesRole (V1.Scripts.Input inputContent) = inputContentUsesRole inputContent
+              inputs :: V1.Scripts.MarloweInput
+              inputs = case AssocMap.lookup (Spending marloweTxOutRef) txInfoRedeemers of
+                Nothing -> traceError "C3j" -- Marlowe redeemer not found.
+                Just (Redeemer bytes) -> case fromBuiltinData bytes of
+                  Just inputs -> inputs
+                  Nothing -> traceError "C3k" -- Marlowe redeemer not decoded.
+          isJust $ find inputUsesRole inputs
```

Comment:
Hm, it seems to me this only checks for the first input, so, if there are several inputs the remaining ones would go without validating and they could have any value?

Reply:
Logic fixed in 10cd2f5284466b3df2062f1c6020838c3b31be46, which requires that _every_ input satisfy the logic
```
choice name matches ∧ role name matches ⇨ choice number matches
```